neps/nep-0413.md

```diff
@@ -0,0 +1,84 @@
+
+## Rationale and alternatives
+
+A possible solution here is to introduce a restricted access key that is only allowed to call a nonexistent function on the owner's account (this will prevent exploitation as anything signed by this key will not execute on-chain). This access key would be used for all ownership verification, eliminating the need to create access keys for every verification request.
+
+###  The problem with a restricted access key
+
+Originally NEAR was designed with the assumption that access keys would only be used to sign transactions. Using access keys for signing arbitrary payloads carries the risk of the signer inadvertently signing surreptitious transactions which the requester could in turn submit to the network.
+
+When using the keys for authentication, another issue that needs to be handled is a replay attack: a malicious app might request the user to authenticate only to authenticate itself to some other app as the user.
+
+If the wallet allows apps to request the user to sign arbitrary data, both attacks would be possible. Thus, some measures are needed to prevent both attacks:
+
+1. To prevent the malicious signing of transactions, a good solution is to make sure that the data to be signed cannot be parsed as a transaction. Having a standard format for the verification payload(different from the transaction format) will solve this.
+
+2. To prevent one app from relaying a signature to a different app, the data to be signed should include the requesting app's identity (e.g. domain name). Again, a well-defined format will facilitate this.
+
+Using an access key restricted in such a way that it cannot send transactions looks like a good temporary solution. It doesn't solve the second problem, though.
+
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead of a query string will improve this (the fragment isn't included in the Referrer), and somehow permitting only a fixed callback URL will improve this even more.
+
+Also, note that NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+Hence, the final solution we are settling on is to user a restricted payload as shown below.
+
+## Specification
+
+### `verifyOwner`
+
+Signs the message and verifies the owner. Message is not sent to blockchain.
+
+```jsx
+interface VerifyOwnerParams {
+  message?: string; // The message requested sign. Defaults to `verify owner` string.
```

Comment:
could we make message mandatory? it simplifies the wallet-side implementation logic, and makes the devs aware of using a personal message to be signed.